File: go/internal/httpserver/handler.go
package httpserver

import (
	"errors"
	"net/http"

	"rebac-primer/internal/authz"
	"rebac-primer/internal/domain"
)

// handler holds the domain operations and implements each route as a method.
// Keeping fields unexported forces callers to go through NewServer.
type handler struct {
	docs domain.DocumentOperations
}

// handleHealth responds with {"ok": true}.
func (h *handler) handleHealth(w http.ResponseWriter, _ *http.Request) {
	writeJSON(w, http.StatusOK, map[string]bool{"ok": true})
}

// handleCreateDocument handles POST /documents.
//
// Request body (JSON):
//
//	{ "id": "...", "title": "...", "body": "...", "workspaceId": "...", "actorId": "..." }
//
// Response: 201 with { "document": {...} }
func (h *handler) handleCreateDocument(w http.ResponseWriter, r *http.Request) {
	var body struct {
		ID          string `json:"id"`
		Title       string `json:"title"`
		Body        string `json:"body"`
		WorkspaceID string `json:"workspaceId"`
		ActorID     string `json:"actorId"`
	}

	if err := readJSON(r, &body); err != nil {
		writeJSON(w, http.StatusBadRequest, errorBody("invalid JSON: "+err.Error()))
		return
	}

	if body.ID == "" || body.Title == "" || body.Body == "" || body.WorkspaceID == "" || body.ActorID == "" {
		writeJSON(w, http.StatusBadRequest, errorBody("id, title, body, workspaceId, and actorId are required"))
		return
	}

	doc, err := h.docs.Create(r.Context(), domain.CreateDocumentInput{
		ID:        body.ID,
		Title:     body.Title,
		Body:      body.Body,
		Workspace: authz.Workspace(body.WorkspaceID),
		Actor:     authz.User(body.ActorID),
	})
	if err != nil {
		h.writeError(w, err)
		return
	}

	writeJSON(w, http.StatusCreated, map[string]any{"document": doc})
}

// handleGetDocument handles GET /documents/{id}?actorId=...
//
// Response: 200 with { "document": {...} }
func (h *handler) handleGetDocument(w http.ResponseWriter, r *http.Request) {
	id := r.PathValue("id")
	actorID := r.URL.Query().Get("actorId")
	if actorID == "" {
		writeJSON(w, http.StatusBadRequest, errorBody("Missing query parameter: actorId"))
		return
	}

	doc, err := h.docs.Read(r.Context(), id, authz.User(actorID))
	if err != nil {
		h.writeError(w, err)
		return
	}

	writeJSON(w, http.StatusOK, map[string]any{"document": doc})
}

// handleUpdateDocument handles PATCH /documents/{id}.
//
// Request body (JSON):
//
//	{ "body": "...", "actorId": "..." }
//
// Response: 200 with { "document": {...} }
func (h *handler) handleUpdateDocument(w http.ResponseWriter, r *http.Request) {
	id := r.PathValue("id")

	var body struct {
		Body    string `json:"body"`
		ActorID string `json:"actorId"`
	}
	if err := readJSON(r, &body); err != nil {
		writeJSON(w, http.StatusBadRequest, errorBody("invalid JSON: "+err.Error()))
		return
	}

	if body.Body == "" || body.ActorID == "" {
		writeJSON(w, http.StatusBadRequest, errorBody("body and actorId are required"))
		return
	}

	doc, err := h.docs.Update(r.Context(), domain.UpdateDocumentInput{
		ID:    id,
		Body:  body.Body,
		Actor: authz.User(body.ActorID),
	})
	if err != nil {
		h.writeError(w, err)
		return
	}

	writeJSON(w, http.StatusOK, map[string]any{"document": doc})
}

// writeError maps domain errors to HTTP status codes.
// This is the Go equivalent of the TS errorResponse function.
func (h *handler) writeError(w http.ResponseWriter, err error) {
	var notFound *domain.DocumentNotFoundError
	if errors.As(err, &notFound) {
		writeJSON(w, http.StatusNotFound, errorBody(err.Error()))
		return
	}

	var forbidden *domain.ForbiddenError
	if errors.As(err, &forbidden) {
		writeJSON(w, http.StatusForbidden, errorBody(err.Error()))
		return
	}

	writeJSON(w, http.StatusBadRequest, errorBody(err.Error()))
}

func errorBody(msg string) map[string]string {
	return map[string]string{"error": msg}
}

File: src/http/handler.ts
import { user, workspace } from "../authz/types.js";
import { DocumentNotFoundError, ForbiddenError } from "../domain/document.js";
import type { DocumentWorkflow } from "../domain/service.js";
import { isJsonObject, stringField, type JsonObject } from "./json.js";

export type HttpRequest = Readonly<{
  method: string;
  path: string;
  query: URLSearchParams;
  body?: unknown;
}>;

export type HttpResponse = Readonly<{
  statusCode: number;
  body: JsonObject;
}>;

export type HttpHandlerConfig = Readonly<{
  documents: DocumentWorkflow;
}>;

export async function handleHttpRequest(
  config: HttpHandlerConfig,
  request: HttpRequest
): Promise<HttpResponse> {
  try {
    return await routeRequest(config.documents, request);
  } catch (error) {
    return errorResponse(error);
  }
}

async function routeRequest(
  documents: DocumentWorkflow,
  request: HttpRequest
): Promise<HttpResponse> {
  if (request.method === "GET" && request.path === "/health") {
    return json(200, { ok: true });
  }

  if (request.method === "POST" && request.path === "/documents") {
    const body = requiredBody(request.body);
    const created = await documents.create({
      id: stringField(body, "id"),
      title: stringField(body, "title"),
      body: stringField(body, "body"),
      workspace: workspace(stringField(body, "workspaceId")),
      actor: user(stringField(body, "actorId"))
    });
    return json(201, { document: created });
  }

  const documentId = matchDocumentPath(request.path);
  if (documentId && request.method === "GET") {
    const actorId = requiredQuery(request.query, "actorId");
    const found = await documents.read(documentId, user(actorId));
    return json(200, { document: found });
  }

  if (documentId && request.method === "PATCH") {
    const body = requiredBody(request.body);
    const updated = await documents.update({
      id: documentId,
      body: stringField(body, "body"),
      actor: user(stringField(body, "actorId"))
    });
    return json(200, { document: updated });
  }

  return json(404, { error: "Route not found" });
}

function requiredBody(body: unknown): JsonObject {
  if (!isJsonObject(body)) {
    throw new Error("Request body must be a JSON object");
  }

  return body;
}

function matchDocumentPath(pathname: string): string | undefined {
  const match = /^\/documents\/([^/]+)$/.exec(pathname);
  return match?.[1];
}

function requiredQuery(query: URLSearchParams, key: string): string {
  const value = query.get(key);
  if (!value) {
    throw new Error(`Missing query parameter: ${key}`);
  }

  return value;
}

function errorResponse(error: unknown): HttpResponse {
  if (error instanceof ForbiddenError) {
    return json(403, { error: error.message });
  }

  if (error instanceof DocumentNotFoundError) {
    return json(404, { error: error.message });
  }

  const message = error instanceof Error ? error.message : "Unknown error";
  return json(400, { error: message });
}

function json(statusCode: number, body: JsonObject): HttpResponse {
  return { statusCode, body };
}
